### DISCUSSION TITLE
Uses of Dynamic allocation on vector(iox::cxx::vector) ?

### DISCUSSION THREAD

**Discussion by givesun** [eclipse-iceoryx/iceoryx#1772] [Q&A] [answered]
I'm interested in iceoryx_hoofs and I have few questions.

I understood the goals of iceoryx_hoofs is to remove dynamic allocation and use fixed static memory.

When I saw the iox::cxx::vector and api emplace_back, There are few point to use dynamic allocation like the followings.

{code}
template <typename T, uint64_t Capacity>
template <typename... Targs>
inline bool vector<T, Capacity>::emplace_back(Targs&&... args) noexcept
{
    if (m_size < Capacity)
    {
        **new (&at(m_size++)) T(std::forward<Targs>(args)...);**
        return true;
    }
    return false;
}
{code}

There is a code to use 'new', I think this is the code using heap.
Is this the right? Or Could you please teach me If I mis-understand?

**Comment by elfenpiff** [comment, marked as answer]
@givesun Here you are mistaken. `new` has in the end to options:
 * allocate memory on the heap and create an object at the heap memory location.
 * create an object at the provided memory location

The difference is subtle but in this case we provide the memory location in the first bracket `(&at(m_size++))` and then call the constructor of `T` with `T(std::forward(args)...)` which results in
```cxx
new (&at(m_size++)) T(std::forward(args)...);
```

If we would allocate this on the heap the call would look like:
```cxx
new T(std::forward(args)...);
```

If you would like to dive deeper into placement new take a look at this article: https://en.cppreference.com/w/cpp/language/new#Placement_new

**Reply by givesun** [reply]
@elfenpiff 
Thanks for your kindly explanation.
I've knew another usage of 'new' operation.
Thanks.
